PHPCompatibility/Tests/Syntax/NewDynamicClassConstantFetchUnitTest.php

```diff
@@ -0,0 +1,130 @@
+<?php
+/**
+ * PHPCompatibility, an external standard for PHP_CodeSniffer.
+ *
+ * @package   PHPCompatibility
+ * @copyright 2012-2020 PHPCompatibility Contributors
+ * @license   https://opensource.org/licenses/LGPL-3.0 LGPL3
+ * @link      https://github.com/PHPCompatibility/PHPCompatibility
+ */
+
+namespace PHPCompatibility\Tests\Syntax;
+
+use PHPCompatibility\Tests\BaseSniffTestCase;
+
+/**
+ * Test the NewDynamicClassConstantFetch sniff.
+ *
+ * @group newDynamicClassConstantFetch
+ * @group syntax
+ *
+ * @covers \PHPCompatibility\Sniffs\Syntax\NewDynamicClassConstantFetchSniff
+ *
+ * @since 10.0.0
+ */
+final class NewDynamicClassConstantFetchUnitTest extends BaseSniffTestCase
+{
+
+    /**
+     * Ensure a warning message when found syntax on earlier versions.
+     *
+     * @dataProvider dataDynamicClassConstantFetch
+     *
+     * @param int $line The line number.
+     *
+     * @return void
+     */
+    public function testDynamicClassConstantFetch($line)
+    {
+        $file = $this->sniffFile(__FILE__, '8.2');
+        $this->assertError($file, $line, 'Dynamic class constant fetch is not available in PHP 8.2 or earlier.');
+    }
+
+    /**
+     * Data provider.
+     *
+     * @return array
+     * @see    testDynamicClassConstantFetch()
+     */
+    public static function dataDynamicClassConstantFetch()
+    {
+        return [
+            [17],
+            [18],
+            [19],
+            [20],
+            [23],
+        ];
+    }
+
+    /**
+     * Ensure a warning message in NOT found syntax on supported versions.
+     *
+     * @dataProvider dataNoViolationsOnValidVersion
+     *
+     * @param int $line The line number.
+     *
+     * @return void
+     */
+    public function testNoViolationsOnValidVersion($line)
+    {
+        $file = $this->sniffFile(__FILE__, '8.3');
+        $this->assertNoViolation($file, $line);
+    }
```

Comment:
With a `testVersion` set to a PHP version on which the syntax is supported by PHP, the sniff should not throw any errors on any line in the file.

In other words, we don't need a data provider, we can just check the whole file in one go.

```suggestion
     * Verify no notices are thrown at all on PHP versions on which the syntax is supported.
     *
     * @return void
     */
    public function testNoViolationsOnValidVersion()
    {
        $file = $this->sniffFile(__FILE__, '8.3');
        $this->assertNoViolation($file);
    }
```